trie/trie.go

```diff
@@ -404,30 +413,78 @@ func (t *Trie) resolveHash(n hashNode, prefix []byte) (node, error) {
 // Hash returns the root hash of the trie. It does not write to the
 // database and can be used even if the trie doesn't have one.
 func (t *Trie) Hash() common.Hash {
-	hash, cached, _ := t.hashRoot(nil, nil)
+	hash, cached, _ := t.hashRoot(nil)
 	t.root = cached
 	return common.BytesToHash(hash.(hashNode))
 }
 
-// Commit writes all nodes to the trie's memory database, tracking the internal
+// oldCommit is the old implementation of Commit, which uses the
+// regular hasher.
+// It writes all nodes to the trie's memory database, tracking the internal
 // and external (for account tries) references.
-func (t *Trie) Commit(onleaf LeafCallback) (root common.Hash, err error) {
+func (t *Trie) oldCommit(onleaf LeafCallback) (root common.Hash, err error) {
 	if t.db == nil {
 		panic("commit called on trie with nil database")
 	}
-	hash, cached, err := t.hashRoot(t.db, onleaf)
+	if t.root == nil {
+		return emptyRoot, nil
+	}
+	h := newHasher(onleaf)
+	defer returnHasherToPool(h)
+	hash, cached, err := h.hash(t.root, t.db, true)
 	if err != nil {
 		return common.Hash{}, err
 	}
 	t.root = cached
 	return common.BytesToHash(hash.(hashNode)), nil
 }
 
-func (t *Trie) hashRoot(db *Database, onleaf LeafCallback) (node, node, error) {
+func (t *Trie) Commit(onleaf LeafCallback) (root common.Hash, err error) {
+	if t.db == nil {
+		panic("commit called on trie with nil database")
+	}
+	if t.root == nil {
+		return emptyRoot, nil
+	}
+	rootHash := t.Hash()
+	h := newCommitter(onleaf)
+	defer returnCommitterToPool(h)
+	var wg sync.WaitGroup
+	if onleaf != nil {
+		wg.Add(1)
+		go h.commitLoop(t.db, &wg)
```

Comment:
When handling WaitGroups, it's generally cleaner to have the Add and the Done at the same place instead of in the called method, keeps things easier to follow and harder to mess up:

```go
wg.Add(1)
go func() {
    defer wg.Done()
    h.commitLoop(t.db)
}
```

Reply:
I'll fix this on https://github.com/ethereum/go-ethereum/pull/20481 instead